### PR TITLE
Documentation for the builders (renamed: build helpers)

### LINKED ISSUE #235858: Documentation: `darwin.builder` does not belong to the `Builder` part
### Describe the bug

`darwin.builder` sets up a **remote builder**. According to the documentation, it is not a [Nix function](https://nixos.org/manual/nix/stable/language/constructs.html#functions) that produce a [derivation](https://nixos.org/manual/nix/stable/language/derivations.html), and its documentation should not be put under that part.

### Steps To Reproduce
1. Visit https://nixos.org/manual/nixpkgs/stable/#sec-darwin-builder

### Metadata
Please run `nix-shell -p nix-info --run "nix-info -m"` and paste the result.

```console
[user@system:~]$ nix-shell -p nix-info --run "nix-info -m"
 - system: `"aarch64-linux"`
 - host os: `Linux 4.14.186-g3813583ccf55`
 - multi-user?: `no`
 - sandbox: `no`
 - version: `nix-env (Nix) 2.11.1`
 - nixpkgs: `/nix/store/k916pxv4jfsxcyjjc86ah552qjy5jwvb-source`
```

### PR DESCRIPTION
**Pull request by ShamrockLee**
## Description of changes

In Nixpkgs, we have several functions that helps defining a derivation. Such function is currently called "builders", and are document in the [Builders](https://nixos.org/manual/nixpkgs/unstable/#part-builders) part in the Nixpkgs manual.

This PR aims to

1. Define and introduce the builders in the Nixpkgs manual.
2. Properly document the function-based attribute recursion (`finalAttrs: { }`) to distinguish it from the recursive attribute (`rec { }`).
3. ~~Document the new approach to define a builder that supports function-based attribute recursion and add such support to existing builder implementations through `lib.extendMkDerivation` and `lib.extendMkDerivationModified` introduced in #234651.~~ (Split out of this PR)
3. Discuss if we wouldike to rename the term "builder" to something else. (Previous discussion: https://github.com/NixOS/nixpkgs/pull/234651#discussion_r1210849842) (candidate: build helper)

This PR also
* Moves Nixpkgs manual section `darwin.linux-builder` from chapter Special build helper to chapter Packages, and emphasize that it is about *remote builder* bootstrap instead of build helpers.
* Moves `doc/build-helpers/packages/` -> `doc/packages/` to make it parallel to `doc/hooks/` and prepare for future migration. (The Packages chapter is mostly not about build helpers.)

Closes #235858

## Things done

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: doc/builders/builders.chapter.md
Comment (fricklerhandwerk) [inline doc/builders/builders.chapter.md:17]: This is quite a mouthful...

How about

```suggestion
## Fixed-point parameters to build helpers {#sec-build-helpers-finalattrs}
```

@roberth @infinisil

**Review by fricklerhandwerk — COMMENTED** [review]
The new documentation is great and the names are terrible. Let the bikeshedfest begin.

File: doc/builders/builders.chapter.md
Comment (infinisil) [reply doc/builders/builders.chapter.md:17]: No real opinion here, either seems fine. I'd avoid "parameters" though

Comment (fricklerhandwerk) [reply doc/builders/builders.chapter.md:17]: Fixed-point build helper arguments?

Comment (ShamrockLee) [reply doc/builders/builders.chapter.md:17]: "Fixed-point arguments" sounds great.

Comment (fricklerhandwerk) [reply doc/builders/builders.chapter.md:17]:
```suggestion
## Fixed-point arguments to build helpers {#sec-build-helpers-finalattrs}
```

**Review by fricklerhandwerk — COMMENTED** [review]
@shamrocklee if you don't have time I could take over the fixups. If there is no further feedback in the next 1-2 weeks we should merge this.

**Comment by ShamrockLee** [comment]
A big wind just blew:

*   doc/builders/packages -> doc/packages (without document changes, just like "languages and frameworks" and "hooks")
*   doc/builders/spacial/darwin-builder.md -> doc/packages/spacial/darwin-builder.md (with document changes)
*   doc/builders -> doc/build-helpers (with document changes)

File: doc/builders/builders.chapter.md
Comment (ShamrockLee) [reply doc/builders/builders.chapter.md]: Thank you for telling me that!

BTW, this part of documentation seems to be a duplication of the function reference of `lib.extendMkDerivation`. I originally write it here just because `lib.customization` was not included in the function reference then.

I think we should either:
1.  Link from here to the function reference, or
2.  Link form the function reference to here.

The first one sounds more sensible to me. It's a pity that we don't have a way to link against the documentation of each function.

**Comment by ShamrockLee** [comment]
Rebase to resolve conflicts with master changes.

**Review by fricklerhandwerk — APPROVED** [review]
This has seen quite a bit of work now, and it's a very strong improvement over the previous state of affairs.

File: doc/builders/builders.chapter.md
Comment (ShamrockLee) [reply doc/builders/builders.chapter.md]: > This should use an `:::{example}` block, but I wouldn't... block on that.

Addressed.

The example block seems to contain rendering bug that (randomly) affects the style of the next few paragraphs (e.g. showing underline when moving the point er across the paragraphs).

File: doc/build-helpers/build-helpers.chapter.md
Comment (infinisil) [inline doc/build-helpers/build-helpers.chapter.md:24]:
```suggestion
As [`stdenv.mkDerivation`'s support of fixed-point arguments](#mkderivation-recursive-attributes) was first included in the Nixpkgs release 22.05, `mkDerivation`-style build helpers don't always come with such support yet.
```

**Comment by ShamrockLee** [comment]
Two examples added to the "Build Helper" introductory chapter.

File: doc/build-helpers/build-helpers.chapter.md
Comment (infinisil) [inline doc/build-helpers/build-helpers.chapter.md:58]: Since this is part of the rendered user docs, this should not reference Nixpkgs-internal files only relevant for contributions. Instead this should just avoid the reference to `all-packages.nix`.

File: doc/build-helpers/build-helpers.chapter.md
Comment (infinisil) [inline doc/build-helpers/build-helpers.chapter.md:98]: Same here

File: doc/build-helpers/build-helpers.chapter.md
Comment (infinisil) [inline doc/build-helpers/build-helpers.chapter.md:102]: Same here, don't link into internal Nixpkgs files from the user docs. If you just need it for an example you could link to a the file on a specific commit though.

File: doc/build-helpers/build-helpers.chapter.md
Comment (infinisil) [inline doc/build-helpers/build-helpers.chapter.md:1]: It's odd to have a section with the same title as its parent section. Instead this could be moved to the parent section directly (into `doc/build-helpers.md`)

File: doc/build-helpers/build-helpers.chapter.md
Comment (infinisil) [inline doc/build-helpers/build-helpers.chapter.md:28]: This function does not actually exist..?

Also, if it does exist, this would be much better in the documentation of the function itself and linking to it from here.

**Review by infinisil — DISMISSED** [review]

File: doc/build-helpers/build-helpers.chapter.md
Comment (ShamrockLee) [reply doc/build-helpers/build-helpers.chapter.md:28]: > Edit: Oh I see it's only part of #234651, then this PR isn't ready yet, I'll mark it as a draft to prevent merging.

Ahh! Thanks a lot for reminding! I simply forgot that #234651 hasn't been merged.

File: doc/build-helpers/build-helpers.chapter.md
Comment (ShamrockLee) [reply doc/build-helpers/build-helpers.chapter.md:1]: I was just imitating `doc/stdenv/stdenv.chapter.md`. At the time this commit was made, there are still `.xml` files, and neither `doc/stdenv.md` nor `doc/builder.md` exists.

File: doc/build-helpers/build-helpers.chapter.md
Comment (ShamrockLee) [reply doc/build-helpers/build-helpers.chapter.md]: Addressed.

File: doc/build-helpers/build-helpers.chapter.md
Comment (ShamrockLee) [reply doc/build-helpers/build-helpers.chapter.md]: Addressed.

File: doc/build-helpers/build-helpers.chapter.md
Comment (ShamrockLee) [reply doc/build-helpers/build-helpers.chapter.md]: I see. The dependent commit hasn't been merged, and I cannot get a stable commit hash. I'll just remove the link.

File: doc/build-helpers/build-helpers.chapter.md
Comment (ShamrockLee) [reply doc/build-helpers/build-helpers.chapter.md]: Now the introduction goes into `doc/build-helpers.md`, and the sections about fixed-point arguments are promoted as a chapter and placed into `doc/build-helpers/fixed-point-arguments.md`.

**Comment by ShamrockLee** [comment]
It is only the last commit that is related to #234651. It's possible to split that commit away if we hope to merge this earlier. (Nixpkgs release 23.11 seems on the way.)

**Comment by fricklerhandwerk** [comment]
Yes, let's split out that last one and merge soon.

**Comment by ShamrockLee** [comment]
> Yes, let's split out that last one and merge soon.

Just dropped it from this feature branch.

File: doc/build-helpers/build-helpers.chapter.md
Comment (ShamrockLee) [reply doc/build-helpers/build-helpers.chapter.md]: The commits that depends on the unmerged PR is now split out.

**Comment by ShamrockLee** [comment]
Rebase to resolve merge conflicts.

**Comment by fricklerhandwerk** [comment]
@Infinisil @roberth @raitobezarius this needs another approval. I'd like to merge that as is, it's a very strong improvement.

**Review by infinisil — APPROVED** [review]
Needs another rebase, but otherwise looking good!

**Comment by ShamrockLee** [comment]
Do we need to add entries in the release note for these changes?

File: doc/builders/builders.chapter.md
Comment (ShamrockLee) [reply doc/builders/builders.chapter.md]: > It's a pity that we don't have a way to link against the documentation of each function.

Update:

We do have a way to link against it. The anchor would be `finction-library-lib.customisation.extendMkDerivation`. I really hope these automatic generation of anchors could be mentioned somewhere in the documentation.

**Pull request merged**
